### DISCUSSION TITLE
✅ [completed] Find MCU root path on connection

### DISCUSSION THREAD

**Discussion by Josverl** [pycom/pymakr-vsc#224] [Ideas]
Then it's also relevant to be clear about the MCUs root folder, as all synch and read/write operations will need to be relative to that. 

Depending on the firmware and hardware the root can be one of:
 - `/flash`  on - Pycom and Lobo derivatives 
 - `/` on Micropython, circuitpython, pycopy 
 - `/sd` on systems that mount an sd card but little flash ( Pyboard 1.1) 

In V1 together with @RalphHogenbirk we added a mpy script to test/detect the writable root for a device.
- [code: shell.js getRootFolder](https://github.com/pycom/pymakr-vsc/blob/680770c502f5042526d3b1cb85706fb19c0951d2/lib/board/shell.js#L69-L88) 
- PR and discussion #38 

V2 should also have a way to determine this, either through a similar method, or by keying of the USB or firmware information, and possibly with a manual device override

**Comment by jakobrosenberg** [comment]
Thanks for pointing this out.

I'm off tomorrow, but I'll have a look as soon as I'm back. It should be easy to fix.

**Reply by Arcadia197** [reply]
Did you manage to get any look on that yet? ESP boards unusable with the current settings for me, as it tries to work with "/flash" everywhere.

**Comment by Arcadia197** [comment]
I had a look at this. Never really worked on something like this before but was motivated.
While [micropython-ctl](https://github.com/metachris/micropython-ctl) has some python scripts including a stat-function, there is no getcwd-function which could be helpful here.

After some checking, the root folder checking could be implemented in Device.js in the _onConnectedHandler right before it checks for the PymakrConf on the board.

A possible version could be something like
```
  getRootFolder() {
    /** @type {import("micropython-ctl-cont/dist-node/src/main").RunScriptOptions} */
    const options = {};
    options.resolveBeforeResult = true;
    const script = "try:\n\t" +
    "import os\n" + 
    "except ImportError:\n\t" + 
    "import uos as os\n" + 
    "try:\n\t" +
    "r = os.getcwd()\n\t" +
    "print(r)\n\t" +
    "except:\n\t" +
    "print('x')\n" + 
    "#";
    const message = this.runScript(script, options);
    /** transscribe message onto root folder, do some string checking or conversion */
  }
```

However, I wasn't able to get this to run
While I wasn't able to provide a push with a working version, maybe it helps you to implement it.

**Comment by jakobrosenberg** [comment]
The latest preview release has experimental support for custom root folders. You can find it [here](https://marketplace.visualstudio.com/items?itemName=pycom.pymakr-preview).

To add a custom path, go into the config and click `Edit in settings.json`:

![image](https://user-images.githubusercontent.com/4153004/168765452-0b0ed30c-3028-4532-ba24-4ac990c2c636.png)

In settings.json, you can then add your own device. You can set the root path for a single device by using its serial number or a range of devices by using a field like `vendor` or `manufacturer`. For a list of available fields, click `f1` and type `List Devices`.

![image](https://user-images.githubusercontent.com/4153004/168766007-99ecd366-809f-4a15-ae34-ed22eb85d734.png)

**Reply by Josverl** [reply]
Works quite decently (once I located the setting.) 
It would be great if that could be part of the 'configure device' menu.
![image](https://user-images.githubusercontent.com/981654/168890123-6cc70190-9707-4f0e-9c47-1657043a3e1a.png)
Pymakr Devices > ... > configure Device > set device root Path

or possibly using a config that recognizes Pycom devices as well as other common devices :
( Assuming all Pycom devices can be recognized by the manufacturer or VID across platforms)
``` json
{
    "pymakr.devices.config": [

        {
            "match": "manufacturer=Pycom.*",
            "field": "rootPath",
            "value": "/flash"
        },
        {
            "match": "manufacturer=.*",
            "field": "rootPath",
            "value": "/"
        },
        {
            "match": "serialNumber=Example1234",
            "field": "rootPath",
            "value": "/custom/path"
        }
    ]
}
```

**Reply by jakobrosenberg** [reply]
> Works quite decently (once I located the setting.) It would be great if that could be part of the 'configure device' menu. ![image]

That could happen, though it would probably be a generic wizard. Something like
```
Pick match:
- path=COM6
- manufacturer=Microchip Technology, Inc.
- serialNumber=Py73bb20
- pnpId=USB\VID_04D8&PID_EF98\PY73BB20
- locationId=Port_#0001.Hub_#0004
- friendlyName=Expansion3 (COM6)
- vendorId=04D8
- productId=EF98

Pick field:
- rootPath

Set value:
<input>
```

Regarding your config, doesn't your `manufacturer=.*` override `manufacturer=Pycom.*`? The entries are processed in the provided order, so the `.*` should run after `Pycom.*`.

**Comment by jakobrosenberg** [comment]
@Arcadia197 that's a great idea. But is CWD always guaranteed to be the root path? What happens if you connect a device where `os.chdir('/scripts')` has been called?

**Reply by Josverl** [reply]
That is indeed why in V1 we found the need for a more complex algorithm to find the topmost (writable) node.
its not perfect, but was deemed 'good enough' at the time.

apart from the javascript wrapping the minimal python code is:
```python 
import os, errno
r = "/flash"
try:
    _ = os.stat(r)
except OSError as e:
    if e.args[0] == errno.ENOENT:
        r = os.getcwd()
finally:
    print(r)
```
This defaults to `/flash` and assumes that os.getcwd() is implemented on the port (Not guarenteed) 

a different variant with less dependencies would be: 
``` python
import os
for r in ["/flash",'/sd','/']:
    try:
        _ = os.stat(r)
        break
    except OSError as e:
        pass
print(r)
```

One important caveat is that running such a script will require interrupting any running application/script on the MCU,
so in V1 this was done just-in-time before uploading/downloading rather than on connecting.

in V2 I propose that it be similarly postponed until needed, and after determining it once the path could be saved, for the session or more permanently.
